### PR TITLE
Introduce skip validation flag for index template API

### PR DESCRIPTION
**Pull request by johtani**
Introduce skip validation param for index template API.
Users can skip validation logic if they set `validate` to `false`.

Add flag and docs

Related #20479

### CODE REVIEW

**Comment by acarstoiu** [comment]
This patch skips validation altogether, which is not correct either. You're accepting any garbage right know as a template.
The best approach is to settle for something in between, i.e. accept the template as an incomplete object, with syntactic validation and maybe semantic, with a field granularity (for mapping).

**Comment by johtani** [comment]
@acarstoiu Thanks for your comment. I agree with your the best approach. 
I thought this PR was first step for avoiding to register a complete template.
We should explore more which properties need to skip validation, i.e. `analyzer` and `search_analyzer`, and maybe `copy_to`.
What do you think?

File: core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java [hunk 1]
```diff
@@ -311,6 +329,12 @@ public PutIndexTemplateRequest source(Map templateSource) {
                 }
             } else if (name.equals("aliases")) {
                 aliases((Map<String, Object>) entry.getValue());
+            } else if (name.equals("validate")) {
+                if (entry.getValue() instanceof Boolean) {
+                    validation((Boolean)entry.getValue());
+                } else if (entry.getValue() instanceof String) {
+                    validation(Booleans.parseBoolean((String)entry.getValue(), DEFAULT_VALIDATION));
+                }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java]: can you throw an exception in a else block?

**Review by jpountz — APPROVED** [review]

**Comment by jpountz** [comment]
I'm afraid the "in-between' approach is very hard. The issue we are having with mappings could well happen with settings too . I think this pull request has the merit to not prevent upgrades. By the way, I think we should consider backporting it to the 2.4 and 5.0 branches.

**Comment by acarstoiu** [comment]
In case you've missed it in #20479, I devised a patch (used right now in production) that could serve as a starting point. It's [here](https://github.com/acarstoiu/elasticsearch/tree/wrong-template-validation).

**Comment by jpountz** [comment]
I don't like just skipping the validation that analyzers exist. On the one hand it means that templates that pass validation could be invalid, and on the other hand there might be more ways that mappings interact with index settings in the future, and skipping the validation that analyzers exist might not be enough anymore.

I would vote to add a flag to skip validation as @johtani suggests and then think about how we can improve templates in the future so that we can better validate them.

**Comment by jpountz** [comment]
> I would vote to add a flag to skip validation as @johtani suggests and then think about how we can improve templates in the future so that we can better validate them.

The other option I'm considering would be to remove template validation entirely for now until we can find a better alternative.

**Comment by jpountz** [comment]
> think about how we can improve templates in the future so that we can better validate them.

#21105

**Comment by acarstoiu** [comment]
Is this patch present in **5.5**? Judgeing by the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html), it's not.

**Comment by javanna** [comment]
@acarstoiu no, this PR was closed without being merged to any branch.

**Pull request closed without merging**